window/src/os/wayland/drag_and_drop.rs

```diff
@@ -0,0 +1,131 @@
+use crate::connection::ConnectionOps;
+use crate::wayland::{read_pipe_with_timeout, WaylandConnection};
+use filedescriptor::{FileDescriptor, Pipe};
+use smithay_client_toolkit as toolkit;
+use std::os::unix::io::AsRawFd;
+use std::path::PathBuf;
+use toolkit::reexports::client::protocol::wl_data_device::Event as DataDeviceEvent;
+use toolkit::reexports::client::protocol::wl_data_offer::WlDataOffer;
+use url::Url;
+use wayland_client::protocol::wl_data_device_manager::DndAction;
+
+#[derive(Default)]
+pub struct DragAndDrop {
+    offer: Option<SurfaceAndOffer>,
+}
+
+struct SurfaceAndOffer {
+    surface_id: u32,
+    offer: WlDataOffer,
+}
+
+struct SurfaceAndPipe {
+    surface_id: u32,
+    read: FileDescriptor,
+}
+
+pub const URI_MIME_TYPE: &str = "text/uri-list";
+
+impl DragAndDrop {
+    /// Takes the current offer, if any, and initiates a receive into a pipe,
+    /// returning that surface and pipe descriptor.
+    fn create_pipe_for_drop(&mut self) -> Option<SurfaceAndPipe> {
+        let SurfaceAndOffer { surface_id, offer } = self.offer.take()?;
+        let pipe = Pipe::new()
+            .map_err(|err| log::error!("Unable to create pipe: {:#}", err))
+            .ok()?;
+        offer.receive(URI_MIME_TYPE.to_string(), pipe.write.as_raw_fd());
+        let read = pipe.read;
+        offer.finish();
+        Some(SurfaceAndPipe {
+            surface_id: surface_id,
+            read,
+        })
+    }
+
+    fn read_paths_from_pipe(read: FileDescriptor) -> Option<Vec<PathBuf>> {
+        read_pipe_with_timeout(read)
+            .map_err(|err| {
+                log::error!("Error while reading pipe from drop result: {:#}", err);
+            })
+            .ok()?
+            .lines()
+            .filter_map(|line| {
+                if line.starts_with('#') || line.trim().is_empty() {
+                    // text/uri-list: Any lines beginning with the '#' character
+                    // are comment lines and are ignored during processing
+                    None
+                } else {
+                    let url = Url::parse(line)
+                        .map_err(|err| {
+                            log::error!(
+                                "Error parsing dropped file line {} as url: {:#}",
+                                line,
+                                err
+                            );
+                        })
+                        .ok()?;
+                    url.to_file_path()
+                        .map_err(|_| {
+                            log::error!(
+                                "Error converting url {} from line {} to pathbuf",
+                                url,
+                                line
+                            );
+                        })
+                        .ok()
+                }
```

Comment:
Can reduce nesting by using early return!

```suggestion
                if line.starts_with('#') || line.trim().is_empty() {
                    // text/uri-list: Any lines beginning with the '#' character
                    // are comment lines and are ignored during processing
                    return None;
                }
                    let url = Url::parse(line)
                        .map_err(|err| {
                            log::error!(
                                "Error parsing dropped file line {} as url: {:#}",
                                line,
                                err
                            );
                        })
                        .ok()?;
                    url.to_file_path()
                        .map_err(|_| {
                            log::error!(
                                "Error converting url {} from line {} to pathbuf",
                                url,
                                line
                            );
                        })
                        .ok()
```

Reply:
Done